File: 2025/python/day5.py
#!/usr/bin/env python3

from aoc import *


def part1(input: str):
    lo = input.splitlines()
    lines = iter(lo)
    fresh = []
    while True:
        line = next(lines)
        if not line:
            break
        (a, b) = line.split('-')
        fresh.append((int(a), int(b)))
    total = 0
    for ingredient in lines:
        for (a, b) in fresh:
            if a <= int(ingredient) <= b:
                total += 1
                break
    return total
    

def part2(input: str):
    fresh = []
    for line in input.splitlines():
        if not line:
            break
        (a, b) = line.split('-')
        fresh.append([int(a), int(b)])

    total = 0
    fresh = sorted(fresh)
    ranges = iter(fresh)
    (a, b) = next(ranges)
    for (x, y) in ranges:
        if a <= x <= b:
            b = max(b, y)
        else:
            total += b - a + 1
            a = x
            b = y
    total += b - a + 1
    return total


if __name__ == '__main__':
    main()

File: 2025/go/day5.go
package main

import (
	"iter"
	"sort"
	"strings"

	"github.com/bpa/aoc"
)

type Range struct {
	Start int
	End   int
}

func day5Part1(input string) int {
	next, _ := iter.Pull(strings.Lines(input))
	fresh := make([]Range, 0, 1024)
	for line, end := next(); end; line, end = next() {
		parts := strings.Split(strings.TrimRight(line, "\r\n"), "-")
		if len(parts) != 2 {
			break
		}
		fresh = append(fresh, Range{
			Start: aoc.ToInt(parts[0]),
			End:   aoc.ToInt(parts[1]),
		})
	}
	total := 0
	for line, end := next(); end; line, end = next() {
		num := aoc.ToInt(line)
		for _, f := range fresh {
			if num >= f.Start && num <= f.End {
				total++
				break
			}
		}
	}
	return total
}

func day5Part2(input string) int {
	next, stop := iter.Pull(strings.Lines(input))
	fresh := make([]Range, 0, 1024)
	for line, end := next(); end; line, end = next() {
		parts := strings.Split(strings.TrimRight(line, "\r\n"), "-")
		if len(parts) != 2 {
			stop()
			break
		}
		fresh = append(fresh, Range{
			Start: aoc.ToInt(parts[0]),
			End:   aoc.ToInt(parts[1]),
		})
	}

	sort.Slice(fresh, func(i, j int) bool {
		return fresh[i].Start < fresh[j].Start
	})

	total := 0
	end := -1
	for _, f := range fresh {
		if f.Start <= end {
			if f.End > end {
				total += f.End - end
				end = f.End
			}
		} else {
			total += f.End - f.Start + 1
			end = f.End
		}
	}

	return total
}

func main() {
	aoc.Main(day5Part1, day5Part2)
}
